File: go/5. MySQL/main.go
package main

import (
	"database/sql"
	"fmt"
	_ "github.com/go-sql-driver/mysql"
	"os"
	"strconv"
	"strings"
	"time"
)

func routine(db *sql.DB, age, id int, ch chan<- string) {
	if age%2 == 0 {
		queryUpdate := fmt.Sprintf("UPDATE users SET salary = salary + 1 WHERE id = %d", id)
		db.Exec(queryUpdate)

		querySelect := fmt.Sprintf("SELECT salary FROM users WHERE id = %d", id)
		var salary string
		db.QueryRow(querySelect).Scan(&salary)
		ch <- fmt.Sprintf("%d: %s", id, salary)
	} else {
		query := fmt.Sprintf("SELECT salary FROM users WHERE age = %d ORDER BY id", age)
		rows, _ := db.Query(query)
		defer rows.Close()

		var salaries []string
		for rows.Next() {
			var salary string
			rows.Scan(&salary)
			salaries = append(salaries, salary)
		}

		ch <- strings.Join(salaries, ", ")
	}
}

func lcg(n, maxValue, seed int) []int {
	timers := make([]int, n)
	timers[0] = seed

	m := 0x7FFFFFFF
	a := 48271
	c := 0
	for i := 1; i < n; i++ {
		timers[i] = (timers[i-1]*a + c) % m
		timers[i-1] = timers[i-1] % maxValue
	}
	timers[n-1] = timers[n-1] % maxValue

	return timers
}

func main() {
	n, _ := strconv.Atoi(os.Args[1])

	ages := lcg(n, 100, 42)
	ids := lcg(n, 10_000, 17)
	ch := make(chan string)
	list := make([]string, n)

	db, _ := sql.Open("mysql", "root:root@/database")
	db.SetConnMaxLifetime(time.Minute * 3)
	db.SetMaxOpenConns(151)
	defer db.Close()

	start := time.Now()
	for i := 0; i < n; i++ {
		go routine(db, ages[i], ids[i], ch)
	}
	for i := 0; i < n; i++ {
		list[i] = <-ch
	}
	fmt.Println(time.Since(start).Microseconds())
}

File: rust/5. MySQL/src/main.rs
use std::error::Error;
use std::time::{Duration, Instant};
use tokio::task::JoinHandle;
use sqlx::mysql::MySqlPool;
use sqlx::mysql::MySqlPoolOptions;

async fn routine(pool: &MySqlPool, age: i32, id: i32) -> Result<String, Box<dyn Error>> {
    if age % 2 == 0 {
        sqlx::query!("UPDATE users SET salary = salary + 1 WHERE id = ( ? )", id)
            .execute(pool)
            .await?;
        let row = sqlx::query!("SELECT salary FROM users WHERE id = ( ? )", id)
            .fetch_one(pool)
            .await?;
        Ok(format!("{}: {}", id, row.salary))
    } else {
		let res = sqlx::query!("SELECT salary FROM users WHERE age = ( ? ) ORDER BY id", age)
            .fetch_all(pool)
            .await?
            .iter()
            .map(|it| it.salary.to_string())
            .collect::<Vec<String>>()
            .join(", ");
		Ok(res)
    }
}

fn lcg(n: usize, max_value: i128, seed: i128) -> Vec<i32> {
    let mut list: Vec<i128> = Vec::with_capacity(n);
    list.push(seed);

    let m = 0x7FFFFFFF;
    let a = 48271;
    let c = 0;
    for i in 1..n {
        list.push((list[i - 1] * a + c) % m);
        list[i - 1] = list[i - 1] % max_value;
    }
    list[n - 1] = list[n - 1] % max_value;

    list.iter().map(|it| *it as i32).collect()
}

#[tokio::main]
async fn main() {
    let n = std::env::args().nth(1).unwrap().parse().unwrap();
    let pool = MySqlPoolOptions::new()
        .max_connections(151)
        .acquire_timeout(Duration::from_secs(180))
        .connect(&std::env::var("DATABASE_URL").unwrap())
        .await
        .unwrap();
    let ages = lcg(n, 100, 42);
    let ids = lcg(n, 10_000, 17);

    let start = Instant::now();
    let handles: Vec<JoinHandle<_>> = (0..n).map(|i| {
        let pool = pool.clone();
        let age = ages[i];
        let id = ids[i];
        tokio::spawn(async move {
            routine(&pool, age, id).await.unwrap()
        })
    }).collect();

    let mut list = Vec::with_capacity(n);
    for handle in handles {
        list.push(handle.await.unwrap());
    }
    _ = list.len();

    println!("{:?}", start.elapsed().as_micros());
}
